### PR TITLE
Ensure redirects work on non-404 pages [ST-916]

### PR DESCRIPTION
**Pull request by tdphillips**
This works. If no one has vehement objections, I'll document everything, add tests, and call this done.

### CODE REVIEW

**Comment by aspidites** [comment]
I still don't get the ticket, so no objections from me. Bonus points for using comprehensions though. :+1:

File: middleware.py [hunk 1]
```diff
@@ -283,4 +284,17 @@ def custom_facets_ops_groups(site_facets):
...
+
+
+class RedirectOverrideMiddleware(object):
+    def process_request(self, request):
+        configuration = get_site_config(request)
+        if configuration.not_found_override.exists():
+            paths = [request.path,
+                     (request.path[:-1] if request.path.endswith('/')
+                      else request.path + '/')]
+            not_found = configuration.not_found_override.filter(
```
Comment (tdphillips) [inline middleware.py]: Breaks if the department that must not be named wants "/path" and "/path/" to go to different locations. If that happens... ¯\\_(ツ)_/¯

**Comment by tdphillips** [comment]
Reiterating, my commit message is completely wrong yet the code does what I said it would in the ticket. :unamused:

File: seo/tests/setup.py [hunk 2]
```diff
@@ -67,7 +67,8 @@ def setUp(self):
 
         self.base_middleware_classes = settings.MIDDLEWARE_CLASSES
         middleware_classes = self.base_middleware_classes + (
-            'wildcard.middleware.WildcardMiddleware', )
+            'wildcard.middleware.WildcardMiddleware',
+            'middleware.RedirectOverrideMiddleware')
```
Comment (tdphillips) [inline seo/tests/setup.py]: Remove this middleware if you wish to confirm this for yourself.

**Comment by aspidites** [comment]
@tdphillips the redirect override middleware is the one to remove for testing locally?

**Comment by tdphillips** [comment]
@aspidites Correct.

**Comment by tdphillips** [comment]
Tests passed locally but are failing on Jenkins. I'll look into it.

**Comment by aspidites** [comment]
@tdphillips, the obligatory annoying question: You're running tests with MySQL locally?

**Comment by tdphillips** [comment]
I did, yes.

File: seo/migrations/0021_auto.py [hunk 3]
```diff
@@ -0,0 +1,424 @@
+# -*- coding: utf-8 -*-
+from south.utils import datetime_utils as datetime
```
Comment (aspidites) [inline seo/migrations/0021_auto.py]: Would you mind renaming this migration to something more descriptive, please?

Comment (tdphillips) [inline seo/migrations/0021_auto.py]: I'll do so. Yay, autonamed migrations.

**Comment by aspidites** [comment]
Given bogus URL A, legit URL B, and redirect URL C, where A doesn't exist, B does, and C is where I want to go:

Without the middleware:
- I can go from A to C
- I can not go from B to C

Which, if I've understood correctly, is the expected (though undesired) behavior. However, with the middleware this behavior is unchanged. where as I would expect to be able to go from B to C.

**Comment by tdphillips** [comment]
@aspidites Don't look too hard into it if you don't remember, but do you recall what the paths for A, B, and C were from your previous comment?

File: middleware.py [hunk 4]
```diff
@@ -246,3 +247,24 @@ def custom_facets_ops_groups(site_facets):
...
+class RedirectOverrideMiddleware(object):
+    """
+    Does a redirect if one is set on the current configuration for the current
+    path. Default behavior was to not do redirects if the page did not result
+    in a 404.
+    """
+    def process_request(self, request):
+        # Configuration is cached. The effect on response time is
+        # hopefully minimal.
+        configuration = get_site_config(request)
```
Comment (JLMcLaugh) [inline middleware.py]: If we're going to start calling get_site_config more than once per request, we should memoize it.

Comment (tdphillips) [inline middleware.py]: What would memoizing get us that cache wouldn't?

https://code.djangoproject.com/ticket/21351#comment:6

Comment (JLMcLaugh) [inline middleware.py]: memoize is local, a call to memcached is fast, but not free. I assume it's microseconds vs milliseconds, but I could be wrong.

Comment (JLMcLaugh) [inline middleware.py]: I realized you were talking about Django's queryset caching, not memcached. You're correct.

**Pull request merged**